source/particles/particle_handler.cc

```diff
@@ -205,66 +224,158 @@ namespace Particles
+        // of it (on the order of 20+ instructions with many
+        // difficult-to-predict branches). Therefore, we simply copy the list
+        // into a new one (keeping alive the possibly large vectors with
+        // particles on cells) into a new container.
+        particle_container sorted_particles;
+        reset_particle_container(&*triangulation, sorted_particles);
+
+        // iterate over cells and insert the entries in the new order
+        for (const auto &cell : triangulation->active_cell_iterators())
+          if (!cell->is_artificial())
+            if (cells_to_particle_cache[cell->active_cell_index()] !=
+                particles.end())
+              {
+                // observe that owned_particles_end was already set to point
+                // into the new sorted_particles array
+                typename particle_container::iterator insert_position =
+                  cell->is_locally_owned() ? owned_particles_end :
+                                             --sorted_particles.end();
+                typename particle_container::iterator new_entry =
+                  sorted_particles.insert(
+                    insert_position, typename particle_container::value_type());
+                new_entry->cell = cell;
+                new_entry->particles =
+                  std::move(cells_to_particle_cache[cell->active_cell_index()]
+                              ->particles);
+              }
+        particles = std::move(sorted_particles);
+
+        // refresh cells_to_particle_cache
+        cells_to_particle_cache.clear();
+        cells_to_particle_cache.resize(triangulation->n_active_cells(),
+                                       particles.end());
+        for (auto it = particles.begin(); it != particles.end(); ++it)
+          if (!it->particles.empty())
+            cells_to_particle_cache[it->cell->active_cell_index()] = it;
+      }
+
+    // Ensure that we did not accidentally modify the anchor entries with
+    // special purpose.
+    Assert(particles.front().cell.state() == IteratorState::past_the_end &&
+             particles.front().particles.empty() &&
+             particles.back().cell.state() == IteratorState::past_the_end &&
+             particles.back().particles.empty() &&
+             owned_particles_end->cell.state() == IteratorState::past_the_end &&
+             owned_particles_end->particles.empty(),
+           ExcInternalError());
+
+#ifdef DEBUG
+    // check that no cache element hits the three anchor states in the list of
+    // particles
+    for (const auto &it : cells_to_particle_cache)
+      Assert(it != particles.begin() && it != owned_particles_end &&
+               it != --(particles.end()),
+             ExcInternalError());
+
+    // check that we only have locally owned particles in the first region of
+    // cells; note that we skip the very first anchor element
+    for (auto it = ++particles.begin(); it != owned_particles_end; ++it)
```

Comment:
I see two possibilities: Either we set the variable to `const` and only set it in a single place by a `const_cast`, or we use some internal functions `owned_particles_begin`, `owned_particles_end`, `ghost_particles_begin`, `ghost_particles_end` that we use everywhere, except for the initialization. This would limit the access confusion (with the three anchors) to a single place where we retrieved them. Opinions?

Reply:
I opted for additional `begin()/end()` functions to be used internally. This should streamline use.